docs/Client.md

```diff
@@ -0,0 +1,1011 @@
+
+Arguments:
+
+* **options** `PipelineOptions`
+* **handler** `(data: PipelineHandlerData) => stream.Readable`
+
+Returns: `stream.Duplex`
+
+#### Parameter: PipelineOptions
+
+Extends: [`RequestOptions`](#parameter-requestoptions)
+
+* **objectMode** `boolean` (optional) - Default: `false` - Set to `true` if the `handler` will return an object stream.
+
+#### Parameter: PipelineHandlerData
+
+* **statusCode** `number`
+* **headers** `IncomingHttpHeaders`
+* **opaque** `unknown`
+* **body** `stream.Readable`
+
+#### Example 1 - Pipeline Echo
+
+```js
+'use strict'
+const { Readable, Writable, PassThrough, pipeline } = require('stream')
+const { createServer } = require('http')
+const { Client } = require('undici')
+
+
+const server = createServer((request, response) => {
+  request.pipe(response)
+})
+
+server.listen(() => {
+  const client = new Client(`http://localhost:${server.address().port}`)
+
+  let res = ''
+
+  pipeline(
+    new Readable({
+      read () {
+        this.push(Buffer.from('undici'))
+        this.push(null)
+      }
+    }),
+    client.pipeline({
+      path: '/',
+      method: 'GET'
+    }, ({ statusCode, headers, body }) => {
+      console.log(`response received ${statusCode}`)
+      console.log('headers', headers)
+      return pipeline(body, new PassThrough(), () => {})
+    }),
+    new Writable({
+      write (chunk, _, callback) {
+        res += chunk.toString()
+        callback()
+      },
+      final (callback) {
+        console.log(`Response pipelined to writable: ${res}`)
+        callback()
+      }
+    }),
+    error => {
+      if (error) {
+        console.error(error)
+      }
+
+      client.close()
+      server.close()
+    }
+  )
+})
+```
+
+### `Client.request(options [, callback])`
+
+Performs a HTTP request.
+
+Non-idempotent requests will not be pipelined in order
+to avoid indirect failures.
+
+Idempotent requests will be automatically retried if
+they fail due to indirect failure from the request
+at the head of the pipeline. This does not apply to
+idempotent requests with a stream request body.
+
+Arguments:
+
+* **options** `RequestOptions`
+* **callback** `(error: Error | null, data: ResponseData) => void` (optional)
+
+Returns: `void | Promise<ResponseData>` - Only returns a `Promise` if no `callback` argument was passed
+
+#### Parameter: `RequestOptions`
+
+Extends: [`DispatchOptions`](#parameter-dispatchoptions)
+
+* **opaque** `unknown` (optional) - Default: `null` - Used for passing through context to `ResponseData`
+* **signal** `AbortSignal | events.EventEmitter | null` (optional) - Default: `null`
+
+The `RequestOptions.method` property should not be value `'CONNECT'`.
+
+#### Parameter: `ResponseData`
+
+* **statusCode** `number`
+* **headers** `http.IncomingHttpHeaders`
```

Comment:
where/how is this declared?

Reply:
I think these (and others such as stream.Readable etc...) should have a link.

Reply:
My reasoning for using this type is that for most node.js http header objects there is atleast some verification that the user is passing in strings, thus its commonly typed as `Record<string, string>`. But, since this parameter directly corresponds to http headers, the Node.js type `http.IncomingHttpHeaders` is useful for saying `headers` is an object that has keys that match this list of http headers, and the values are also strings. Plus, it supports custom headers as well. 

TLDR: whenever I use http headers and typescript I rely on that Node.js http headers type and so I thought it would be good to include here as well since node is already a dependency. I'd say we need more TS user input to change this to something more general like `Record<string, string>` or just `object`

Reply:
> I think these (and others such as stream.Readable etc...) should have a link.

Yes i agree. still figuring out the best way to represent that in these sections. I'm thinking since you can't embed a markdown link within the code block itself, then I will include a `Refs: ` section which is a list of links for types mentioned in the relative definition block. Will contribute those changes shortly to get your feedback